### DISCUSSION TITLE
Best Practice when Accessing a Riverpod Provider within a Repository Class

### DISCUSSION THREAD

**Discussion by bmmcc4** [rrousselGit/riverpod#2662] [Q&A]
Hello,

My Flutter app uses Riverpod for state management and the [AppAuth](https://pub.dev/packages/flutter_appauth) package to utilize oAuth for authentication. I've created an AppAuth provider to initialize FlutterAppAuth and access its methods within my AuthRepository class. My question is what is the best practice to access the AppAuth methods within my AuthRepository. I've come up with two options below. 

1. Dependency injection in AuthRepository
2. Pass provider ref to AuthRepository

I've read through the documentation and both options appear to follow standards, as passing a ref (option 2) is allowed. However, option 1 makes use of ref.watch which appears to be recommended over ref.read (option 2). But, in the documentation it states to use ref.read when calling a method that will update state (option 2). Finally, since my appAuthProvider is just providing an instance of FlutterAppAuth I don't believe it will ever change, therefore, my authRepositoryProvider would never rebuild with it watching my appAuthProvider (option 1). So, my thinking is the ref.watch is unnecessary.

Below are the two options laid out in code.
Thank you!

```
// Option 1 - Dependency injection in AuthRepository
// AppAuth provider, access to AppAuth methods
final appAuthProvider = Provider<FlutterAppAuth>(
  (ref) => const FlutterAppAuth(),
);

// Auth repository provider, used in StateNotifier to access repository methods
final authRepositoryProvider = Provider<AuthRepository>((ref) {
  final appAuth = ref.watch(appAuthProvider);
  return AuthRepository(appAuth);
});

// Auth repository, use AppAuth methods
class AuthRepository {
  final FlutterAppAuth _appAuth; // Dependency

  const AuthRepository(
    this._appAuth,
  );

  Future<bool> signIn() async {
    // Call method from appAuth dependency
    final authResult = await _appAuth.authorizeAndExchangeCode(
      AuthorizationTokenRequest(
        '<client_id>',
        '<redirect_url>',
        discoveryUrl: '<discovery_url>',
        scopes: ['openid', 'profile', 'email', 'offline_access', 'api'],
      ),
    );

    if (authResult != null) {
      return true;
    }

    return false;
  }
}
```

```
// Option 2 - Pass provider ref to AuthRepository
// AppAuth provider, access to AppAuth methods
final appAuthProvider = Provider<FlutterAppAuth>(
  (ref) => const FlutterAppAuth(),
);

// Auth repository provider, used in StateNotifier to access repository methods
final authRepositoryProvider = Provider<AuthRepository>(
  (ref) => AuthRepository(ref),
);

// Auth repository, use AppAuth methods
class AuthRepository {
  final Ref _ref; // Pass Ref

  const AuthRepository(
    this._ref,
  );

  Future<bool> signIn() async {
    // Read AppAuth provider directly to call method
    final authResult = await _ref.read(appAuthProvider).authorizeAndExchangeCode(
          AuthorizationTokenRequest(
            '<client_id>',
            '<redirect_url>',
            discoveryUrl: '<discovery_url>',
            scopes: ['openid', 'profile', 'email', 'offline_access', 'api'],
          ),
        );

    if (authResult != null) {
      return true;
    }

    return false;
  }
}
```

**Comment by est7** [comment]
Can someone help answer this, I'm also confused and don't know which way to use

**Comment by AhmedLSayed9** [comment]
Both ways are okay.

Using `ref.watch` in the provider body will create a new instance of `AuthRepository` when `appAuthProvider` state changes. So, If you've some logic in your constructor body, it'll run again.

I prefer passing `Ref` approach as using `ref.watch` approach has a small drawback:
If another asynchronous code run at the same method before using your dependency `_appAuth` and your dependency has updated (overridden with new class if it's a provider) at that moment, you'll be using the old dependency `_appAuth` (it's very rare to happen):
```dart
class AuthRepository {
  const AuthRepository(
    this._appAuth,
  );

  final FlutterAppAuth _appAuth; // Dependency
  
  Future<bool> signIn() async {
    await someAsyncWork();  // `appAuthProvider` has overridden with new value while this is running

    // Call method from appAuth dependency
    final authResult = await _appAuth.authorizeAndExchangeCode(...), //This will use the old `appAuthProvider`

    if (authResult != null) {
      return true;
    }
    return false;
  }
}
```

Conclusion: passing ref will make sure you use the most updated state. and If you don't want to repeat `ref.read(..)` when used in multiple methods, you can use a getter for it:
```dart
class AuthRepository {
  const AuthRepository(
    this._ref,
  );

  final Ref _ref; // Pass Ref

  FlutterAppAuth get _appAuth => ref.read(appAuthProvider);
```

**Reply by MartinBergerDX** [reply]
Don't go passing `Ref` around. It used to be the thing, but as of Riverpod 3 I started getting so many errors related to disposed provider `Ref`s it was crazy.

`Ref` is tied to the lifecycle of the provider:

```
final appAuthProvider = Provider<FlutterAppAuth>(
  (ref) => const FlutterAppAuth(), // ref is valid for the lifetime of this provider instance, not inside FlutterAppAuth if you pass it in, if you obtained Provider<FlutterAppAuth> via ref.read
);
```

So if you obtain an object from a provider via ref.read and pass it a Ref, that object now holds a Ref that can (via ref.read) outlive the provider scope. After ref.read, nothing is holding a dependency on the provider that created that Ref.

This is a huge problem if you do async work and then use ref to obtain something from another provider when you get data back — e.g. instantiate a family provider after you obtain an upload URL.

You will get errors like:

```
I/flutter ( 4354): Cannot use the Ref of uploadOnboardingDocumentProvider((applicationId: ..., documentType: MotorInsuranceCertificate, fileName: ..., sourcePath: ... after it has been disposed. This typically happens if:
I/flutter ( 4354): - A provider rebuilt, but the previous "build" was still pending and is still performing operations.
I/flutter ( 4354):   You should therefore either use `ref.onDispose` to cancel pending work, or
I/flutter ( 4354):   check `ref.mounted` after async gaps or anything that could invalidate the provider.
I/flutter ( 4354): - You tried to use Ref inside `onDispose` or other life-cycles.
I/flutter ( 4354):   This is not supported, as the provider is already being disposed.
```


It is difficult tracking causes of these errors, especially if you use Tasks.

Refs should not leave their scopes. Rely on ref.watch in the provider factory for stable one-off deps — for something like FlutterAppAuth, that won't meaningfully rebuild the parent.

Ref is fine inside @riverpod notifiers/providers, as long as it stays in that scope and you respect ref.mounted after async gaps.

They should really improve this, while I love Riverpod, I have spent dozens of hours fighting these Ref issues. Trick is in using ref.watch vs ref.read . Don't pass Ref around, don't use ref.keepAlive etc...

TL;DR: go with option 1. Don't pass Ref into repositories/services, otherwise you may end up relying on ref.keepAlive(), which is also bad because of unexpected usages/lifetimes.
